### ISSUE TITLE
References on Type

### ISSUE THREAD

**Issue by tgockel** [OAI/OpenAPI-Specification#2024]
Let's say I have an API object that looks like this:

```
{
  "foo": "192.168.1.20"
}
```

Now, I would like `foo` to take on a default value specific to this object (let's say `"127.0.0.1"`) and I would like things to still be validated as `NetworkAddressIPv4` types. This is fairly expressible in a number of programming languages. For example, in Kotlin, I would say something like this:

```
data class MyThing(foo: NetworkAddressIPv4 = NetworkAddressIPv4("127.0.0.1"))
```

Unfortunately, I can't seem to figure out how to express this to OpenAPI. What I would expect is:

```
MyThing:
  type: object
  properties:
    foo:
      type: { $ref: '#/components/schemas/NetworkAddressIPv4' }
      default: "127.0.0.1"
```

The problem is that this is a validation error because [OpenAPI schema objects](https://swagger.io/specification/#schemaObject) type field can only be a string and cannot be a [`Reference`](https://github.com/OAI/OpenAPI-Specification/blob/master/schemas/v3.0/schema.yaml#L260). The appears to make doing things like adding `description` or `default` fields to properties whose type references an existing schema impossible.

This seems to be a limitation of and expected behavior of JSON Schema. Per [JSON Schema Structuring](https://json-schema.org/understanding-json-schema/structuring.html) (emphasis mine):

> We can then refer to this schema snippet from elsewhere using the $ref keyword. The easiest way to describe $ref is that it gets logically replaced with the thing that it points to. So, to refer to the above, we would include:
>
> `{ "$ref": "#/definitions/address" }`
>
> *This can be used anywhere a schema is expected.* You will always use $ref as the only key in an object: any other keys you put there will be ignored by the validator.

Of course, `type` does not expect a schema, only a string. But why not allow a `$ref` here? If not, is there some other way to address this use case?

**Comment by tgockel** [comment]
The workaround I use is this:

```
MyThing:
  type: object
  properties:
    foo:
      $ref: '#/components/schemas/NetworkAddressIPv4'
      default: "127.0.0.1"
```

This passes validation by tools and is Good Enough for a human to figure out what is going on. Unfortunately, since other keys are ignored, the `default` is not respected by any of the tooling.

**Comment by slavcodev** [comment]
Try this one to use `default` with `$ref`
```
MyThing:
  type: object
  properties:
    foo:
      allOf:
        - $ref: '#/components/schemas/NetworkAddressIPv4'
        - default: "127.0.0.1"
```

**Comment by handrews** [comment]
@tgockel yes this is a current limitation, and @slavcodev 's solution is correct.

The limitation will be removed in OAS 3.1 per #1977

**Comment by handrews** [comment]
@webron @darrelmiller @MikeRalphson I think this can be closed as it's been answered with workaround for 3.0 and solved permanently in 3.1?

**Comment by webron** [comment]
Yup.
